components/psst/browser/content/psst_scripts_result_handler.cc

```diff
@@ -0,0 +1,380 @@
+    ResetContext();
+    return;
+  }
+
+  const auto* psst = value.GetDict().FindDict("psst");
+  if (!psst) {
+    ResetContext();
+    return;
+  }
+
+  if (const auto percent = psst->FindDouble("progress")) {
+    delegate_->SetProgressValue(*percent);
+  }
+
+  const auto applied_list =
+      ParseAppliedList(delegate_.get(), psst->FindList("applied"));
+  const auto errors_list =
+      ParseErrors(delegate_.get(), psst->FindDict("errors"));
+  if (const auto result = value.GetDict().FindBool("result");
+      !result || !result.value()) {
+    return;
+  }
+
+  delegate_->SetCompletedView(applied_list, errors_list);
+
+  ResetContext();
+}
+
+void PsstScriptsHandlerImpl::InsertUserScript(
+    const std::optional<MatchedRule>& rule) {
+  if (!rule) {
+    return;
+  }
+
+  InsertScriptInPage(rule->UserScript(), std::nullopt /* no params */,
+                     base::BindOnce(&PsstScriptsHandlerImpl::OnUserScriptResult,
+                                    weak_factory_.GetWeakPtr(), rule.value()));
+}
+
+void PsstScriptsHandlerImpl::InsertPolicyScript(
+    const std::optional<MatchedRule>& rule) {
+  if (!rule) {
+    ResetContext();
+    return;
+  }
+
+  const auto settings_for_site =
+      GetPsstSettings(context_->GetUserId(), context_->GetRuleName(), prefs_);
+
+  if (!settings_for_site || settings_for_site->consent_status == kBlock) {
+    ResetContext();
+    return;
+  }
+
+  std::optional<base::Value> params;
+  base::Value::Dict dict;
+  params = base::Value(std::move(dict));
+  PrepareParametersForPolicyExecution(params, settings_for_site->urls_to_skip,
+                                      false);
+
+  InsertScriptInPage(
+      rule->PolicyScript(), std::move(params),
+      base::BindOnce(&PsstScriptsHandlerImpl::OnPolicyScriptResult,
+                     weak_factory_.GetWeakPtr(), rule.value()));
+}
+
+bool PsstScriptsHandlerImpl::TryToLoadContext(const MatchedRule& rule,
+                                              base::Value& script_result) {
+  context_ = PsstOperationContext::LoadContext(script_result, rule);
+
+  if (!context_ || !context_->IsValid()) {
+    ResetContext();
+    return false;
+  }
+
+  return true;
+}
+
+void PsstScriptsHandlerImpl::ResetContext() {
+  context_ = nullptr;
+}
+
+void PsstScriptsHandlerImpl::DisablePsst() {
+  SetEnablePsstFlag(prefs_, false);
+  ResetContext();
+}
+
+PsstDialogDelegate* PsstScriptsHandlerImpl::GetPsstDialogDelegate() {
+  return delegate_.get();
+}
+
+void PsstScriptsHandlerImpl::InsertScriptInPage(
+    const std::string& script,
+    std::optional<base::Value> value,
+    InsertScriptInPageCallback cb) {
+  content::RenderFrameHost* render_frame_host =
+      content::RenderFrameHost::FromID(render_frame_host_id_);
+
+  // Check if render_frame_host is still valid and if starting rfh is the same.
+  if (!render_frame_host ||
+      render_frame_host_id_ !=
+          web_contents_->GetPrimaryMainFrame()->GetGlobalId()) {
+    return;
```

Comment:
Q: Do you need to run the callback with an empty value here to reset the context?